fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializedViewOptimizer.java

```diff
@@ -57,42 +65,103 @@ public List<ColumnRefOperator> getOutputExpressions() {
+        partitionExpr.collect(SlotRef.class, partitionSlotRefs);
+        Preconditions.checkState(partitionSlotRefs.size() == 1);
+        ColumnRefFactory refFactory = new ColumnRefFactory();
+        ColumnRefOperator partitionColumnRef = refFactory.create(
+                partitionSlotRefs.get(0).getColumnName(), partitionColumn.getType(), partitionColumn.isAllowNull());
+        ExpressionMapping expressionMapping =
+                new ExpressionMapping(new Scope(RelationId.anonymous(), new RelationFields()),
+                        Lists.newArrayList());
+        expressionMapping.put(partitionSlotRefs.get(0), partitionColumnRef);
+        ScalarOperator partitionScalarForScan =
+                SqlToScalarOperatorTranslator.translate(partitionExpr, expressionMapping, refFactory);
+        List<ScalarOperator> partitionPredicatesForScan = convertRanges(partitionScalarForScan, mvRanges);
+        return Utils.compoundOr(partitionPredicatesForScan);
+    }
+
+    // try to get partial partition predicate of partitioned mv.
     // for example, mv1 has two partition: p1:[2022-01-01, 2022-01-02), p2:[2022-01-02, 2022-01-03).
     // p1 is updated, p2 is outdated.
     // mv1's base partition table is t1, partition column is k1.
     // then this function will add predicate: k1 >= "2022-01-01" and k1 < "2022-01-02" to scan node of t1
-    public void getPartialPartitionPredicate(MaterializedView mv, ColumnRefFactory columnRefFactory,
-                                             Set<String> partitionsToRefresh, OptExpression mvPlan) {
-        // to support partial partition rewrite
-        PartitionInfo partitionInfo = mv.getPartitionInfo();
-        if (!(partitionInfo instanceof ExpressionRangePartitionInfo)) {
-            return;
+    private boolean updatePartitionByScanPredicate(MaterializedView mv, OptExpression mvPlan) {
+        // convert finalRanges into ScalarOperator
+        Pair<Table, Column> partitionTableAndColumns = mv.getPartitionTableAndColumn();
+        if (partitionTableAndColumns == null) {
+            return false;
         }
-        ExpressionRangePartitionInfo exprPartitionInfo = (ExpressionRangePartitionInfo) partitionInfo;
-        if (partitionsToRefresh.isEmpty()) {
-            // all partitions are uptodate, do not add filter
-            return;
+        OlapTable partitionByTable = (OlapTable) partitionTableAndColumns.first;
+        Set<String> modifiedPartitionNames = mv.getUpdatedPartitionNamesOfTable(partitionByTable);
+        List<Range<PartitionKey>> baseTableRanges = getUptodatedPartitionRange(partitionByTable, modifiedPartitionNames);
+        mvRanges = getUptodatedPartitionRange(mv, partitionNamesToRefresh);
+        List<Range<PartitionKey>> finalRanges = Lists.newArrayList();
+        for (Range<PartitionKey> range : baseTableRanges) {
+            if (mvRanges.stream().anyMatch(mvRange -> mvRange.encloses(range))) {
+                finalRanges.add(range);
```

Comment:
Could you plz add more comment for this finalRanges? Or give a suitable name I can understand easily?